vm/devices/storage/disklayer_sqlite/src/lib.rs

```diff
@@ -0,0 +1,617 @@
+    }
+
+    fn optimal_unmap_sectors(&self) -> u32 {
+        1
+    }
+}
+
+impl WriteNoOverwrite for SqliteDiskLayer {
+    async fn write_no_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> Result<(), DiskError> {
+        self.write_maybe_overwrite(buffers, sector, false).await
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum SectorKind {
+    AllZero,
+    AllOne,
+    Data(Vec<u8>),
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn read_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    start_sector: u64,
+    end_sector: u64,
+) -> Result<Vec<(u64, SectorKind)>, rusqlite::Error> {
+    let conn = conn.lock();
+
+    let mut select_stmt = conn.prepare_cached(
+        "SELECT sector, data
+        FROM sectors
+        WHERE sector >= ? AND sector < ?
+        ORDER BY sector ASC",
+    )?;
+    let mut rows = select_stmt.query(rusqlite::params![start_sector, end_sector])?;
+
+    // reuse the same blob handle across all reads.
+    //
+    // this handle isn't easy to persist across multiple calls to
+    // `read_sectors`, due to the lifetime restrictions imposed by using a
+    // `Arc<Mutex<Connection>>`.
+    let Some(first_row) = rows.next()? else {
+        return Ok(Vec::new());
+    };
+    let mut sector: u64 = first_row.get(0)?;
+    let mut blob = conn.blob_open(
+        rusqlite::DatabaseName::Main,
+        "sectors",
+        "data",
+        sector as i64,
+        true,
+    )?;
+
+    let mut buf = vec![0; sector_size as usize];
+    let mut res = Vec::new();
+    loop {
+        let data = match blob.len() {
+            0 => SectorKind::AllZero,
+            1 => SectorKind::AllOne,
+            len => {
+                if len != sector_size as usize {
+                    return Err(rusqlite::Error::BlobSizeError);
+                }
+                blob.read_at(&mut buf, 0)?;
+                SectorKind::Data(buf.clone())
+            }
+        };
+
+        res.push((sector, data));
+
+        let Some(next_row) = rows.next()? else {
+            break;
+        };
+
+        sector = next_row.get(0)?;
+        blob.reopen(sector as i64)?;
+    }
+
+    Ok(res)
+}
+
+// FUTURE: pass RequestBuffers into this function directly, and avoid the alloc
+// overhead.
+fn write_sectors(
+    conn: Arc<Mutex<Connection>>,
+    sector_size: u32,
+    mut sector: u64,
+    buf: Vec<u8>,
+    overwrite: bool,
+) -> Result<(), rusqlite::Error> {
+    let mut conn = conn.lock();
+
+    let tx = conn.transaction()?;
+    {
+        let mut stmt = if overwrite {
+            tx.prepare_cached("INSERT OR REPLACE INTO sectors (sector, data) VALUES (?, ?)")?
+        } else {
+            tx.prepare_cached("INSERT OR IGNORE INTO sectors (sector, data) VALUES (?, ?)")?
+        };
+
+        let chunks = buf.chunks_exact(sector_size as usize);
+        if !chunks.remainder().is_empty() {
```

Comment:
Can we assert on this?

Reply:
Do we guarantee that as part of the current trait contract? Or is this something that is true today (and will be true in the future), but isn't currently explicitly claimed anywhere?

Reply:
We should write down in the traits (and in particular on `Disk`) that the caller must pass a buffer with an integer number of sectors. I don't think we validate it anywhere, but all clients should be doing this already. If they're not, then they're not getting the result they expect, so we want to find those bugs anyway.